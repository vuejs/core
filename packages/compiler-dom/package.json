{
  "name": "@vue/compiler-dom",
  "version": "3.5.13",
  "description": "@vue/compiler-dom",
  "main": "index.js",
  "module": "dist/compiler-dom.esm-bundler.js",
  "types": "dist/compiler-dom.d.ts",
  "unpkg": "dist/compiler-dom.global.js",
  "jsdelivr": "dist/compiler-dom.global.js",
  "files": [
    "index.js",
    "dist"
  ],
  "exports": {
    ".": {
      "types": "./dist/compiler-dom.d.ts",
      "node": {
        "production": "./dist/compiler-dom.cjs.prod.js",
        "development": "./dist/compiler-dom.cjs.js",
        "default": "./index.js"
      },
      "module": "./dist/compiler-dom.esm-bundler.js",
      "import": "./dist/compiler-dom.esm-bundler.js",
      "require": "./index.js"
    },
    "./*": "./*"
  },
  "sideEffects": false,
  "buildOptions": {
    "name": "VueCompilerDOM",
    "compat": true,
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "global"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuejs/core.git",
    "directory": "packages/compiler-dom"
  },
  "keywords": [
    "vue"
  ],
  "author": "Evan You",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vuejs/core/issues"
  },
  "homepage": "https://github.com/vuejs/core/tree/main/packages/compiler-dom#readme",
  "dependencies": {
    "@vue/shared": "workspace:*",
    "@vue/compiler-core": "workspace:*"
  }
}
