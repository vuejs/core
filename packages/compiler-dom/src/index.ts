import {
  type CodegenResult,
  type CompilerOptions,
  type DirectiveTransform,
  type NodeTransform,
  type ParserOptions,
  type RootNode,
  baseCompile,
  baseParse,
  noopDirectiveTransform,
} from '@vue/compiler-core'
import { parserOptions } from './parserOptions'
import { transformStyle } from './transforms/transformStyle'
import { transformVHtml } from './transforms/vHtml'
import { transformVText } from './transforms/vText'
import { transformModel } from './transforms/vModel'
import { transformOn } from './transforms/vOn'
import { transformShow } from './transforms/vShow'
import { transformTransition } from './transforms/Transition'
import { stringifyStatic } from './transforms/stringifyStatic'
import { ignoreSideEffectTags } from './transforms/ignoreSideEffectTags'
import { extend } from '@vue/shared'

export { parserOptions }

export const DOMNodeTransforms: NodeTransform[] = [
  transformStyle,
  ...(__DEV__ ? [transformTransition] : []),
]

export const DOMDirectiveTransforms: Record<string, DirectiveTransform> = {
  cloak: noopDirectiveTransform,
  html: transformVHtml,
  text: transformVText,
  model: transformModel, // override compiler-core
  on: transformOn, // override compiler-core
  show: transformShow,
}

export function compile(
  src: string | RootNode,
  options: CompilerOptions = {},
): CodegenResult {
  return baseCompile(
    src,
    extend({}, parserOptions, options, {
      nodeTransforms: [
        // ignore <script> and <tag>
        // this is not put inside DOMNodeTransforms because that list is used
        // by compiler-ssr to generate vnode fallback branches
        ignoreSideEffectTags,
        ...DOMNodeTransforms,
        ...(options.nodeTransforms || []),
      ],
      directiveTransforms: extend(
        {},
        DOMDirectiveTransforms,
        options.directiveTransforms || {},
      ),
      transformHoist: __BROWSER__ ? null : stringifyStatic,
    }),
  )
}

export function parse(template: string, options: ParserOptions = {}): RootNode {
  return baseParse(template, extend({}, parserOptions, options))
}

export * from './runtimeHelpers'
export { transformStyle } from './transforms/transformStyle'
export {
  createDOMCompilerError,
  DOMErrorCodes,
  DOMErrorMessages,
} from './errors'
export * from '@vue/compiler-core'
