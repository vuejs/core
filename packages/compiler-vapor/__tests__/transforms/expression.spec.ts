import { BindingTypes } from '@vue/compiler-dom'
import {
  transformChildren,
  transformElement,
  transformText,
  transformVBind,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithExpression = makeCompile({
  nodeTransforms: [transformElement, transformChildren, transformText],
  directiveTransforms: { bind: transformVBind },
})

describe('compiler: expression', () => {
  test('basic', () => {
    const { code } = compileWithExpression(`{{ a }}`)
    expect(code).toMatchSnapshot()
    expect(code).contains(`ctx.a`)
  })

  test('props', () => {
    const { code } = compileWithExpression(`{{ foo }}`, {
      bindingMetadata: { foo: BindingTypes.PROPS },
    })
    expect(code).toMatchSnapshot()
    expect(code).contains(`$props.foo`)
  })

  test('props aliased', () => {
    const { code } = compileWithExpression(`{{ foo }}`, {
      bindingMetadata: {
        foo: BindingTypes.PROPS_ALIASED,
        __propsAliases: { foo: 'bar' } as any,
      },
    })
    expect(code).toMatchSnapshot()
    expect(code).contains(`$props['bar']`)
  })

  test('update expression', () => {
    const { code } = compileWithExpression(`
      <div :id="String(foo.id++)" :foo="foo" :bar="bar++">
        {{ String(foo.id++) }} {{ foo }} {{ bar }}
      </div>
    `)
    expect(code).toMatchSnapshot()
    expect(code).contains(`_String(_foo.id++)`)
  })

  test('empty interpolation', () => {
    const { code } = compileWithExpression(`{{}}`)
    const { code: code2 } = compileWithExpression(`{{ }}`)
    const { code: code3 } = compileWithExpression(`<div>{{ }}</div>`)
    const { code: code4 } = compileWithExpression(`<div>{{ foo }}{{ }}</div>`)

    expect(code).toMatchSnapshot()
    expect(code).not.toContain(`_toDisplayString`)
    expect(code).not.toContain(`_setText`)

    expect(code2).toMatchSnapshot()
    expect(code2).not.toContain(`_toDisplayString`)
    expect(code2).not.toContain(`_setText`)

    expect(code3).toMatchSnapshot()
    expect(code3).not.toContain(`_toDisplayString`)
    expect(code3).not.toContain(`_setText`)

    expect(code4).toMatchSnapshot()
  })
})
