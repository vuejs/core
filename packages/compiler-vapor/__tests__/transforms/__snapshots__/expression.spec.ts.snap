// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: expression > basic 1`] = `
"import { toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const n0 = t0()
  _renderEffect(() => _setText(n0, _toDisplayString(_ctx.a)))
  return n0
}"
`;

exports[`compiler: expression > empty interpolation 1`] = `
"import { template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`compiler: expression > empty interpolation 2`] = `
"import { template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`compiler: expression > empty interpolation 3`] = `
"import { template as _template } from 'vue';
const t0 = _template("<div></div>", true)

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`compiler: expression > empty interpolation 4`] = `
"import { child as _child, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, template as _template } from 'vue';
const t0 = _template("<div> </div>", true)

export function render(_ctx) {
  const n1 = t0()
  const n0 = _child(n1)
  const x1 = _child(n1)
  _renderEffect(() => {
    const _foo = _ctx.foo
    _setText(n0, _toDisplayString(_foo))
    _setText(x1, _toDisplayString(_foo))
  })
  return n1
}"
`;

exports[`compiler: expression > props 1`] = `
"import { toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = t0()
  _renderEffect(() => _setText(n0, _toDisplayString($props.foo)))
  return n0
}"
`;

exports[`compiler: expression > props aliased 1`] = `
"import { toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = t0()
  _renderEffect(() => _setText(n0, _toDisplayString($props['bar'])))
  return n0
}"
`;

exports[`compiler: expression > update expression 1`] = `
"import { child as _child, setProp as _setProp, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, template as _template } from 'vue';
const t0 = _template("<div> </div>", true)

export function render(_ctx) {
  const n1 = t0()
  const n0 = _child(n1)
  const x1 = _child(n1)
  _renderEffect(() => {
    const _String = String
    const _foo = _ctx.foo
    _setProp(n1, "id", _String(_foo.id++))
    _setProp(n1, "foo", _foo)
    _setProp(n1, "bar", _ctx.bar++)
    _setText(n0, _toDisplayString(_String(_foo.id++)) + " " + _toDisplayString(_foo) + " " + _toDisplayString(_ctx.bar))
    _setText(x1, _toDisplayString(_String(_foo.id++)) + " " + _toDisplayString(_foo) + " " + _toDisplayString(_ctx.bar))
  })
  return n1
}"
`;
